### PR TITLE
[BugFix] fix partial partition rewrite bug

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #13815 

## Problem Summary(Required) ：

fix partial partition rewrite bug by adding partition range predicate to partition by table
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by ABingHuang** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 72 / 78 (92.31%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|46|52|88.46%|[51, 71, 78, 88, 89, 131]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java [hunk 1]
```diff
@@ -62,85 +63,90 @@ public List<ColumnRefOperator> getOutputExpressions() {
...
+            scanOperator.setPredicate(newPredicate);
         }
+        return true;
+    }
 
-        Set<Long> outdatePartitionIds = Sets.newHashSet();
-        for (Partition partition : mv.getPartitions()) {
-            if (partitionsToRefresh.contains(partition.getName())) {
-                outdatePartitionIds.add(partition.getId());
+    private List<Range<PartitionKey>> getUptodatePartitionRangeForTable(OlapTable partitionByTable,
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:104]: Lastest ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java [hunk 2]
```diff
@@ -62,85 +63,90 @@ public List<ColumnRefOperator> getOutputExpressions() {
...
+        for (OptExpression scanExpr : scanExprs) {
+            LogicalScanOperator scanOperator = (LogicalScanOperator) scanExpr.getOp();
+            Table scanTable = scanOperator.getTable();
+            if ((scanTable.isLocalTable() && !scanTable.equals(partitionTableAndColumns.first))
+                    || (!scanTable.isLocalTable()) && !scanTable.getTableIdentifier().equals(
+                    partitionTableAndColumns.first.getTableIdentifier())) {
+                continue;
+            }
+            Optional<Column> columnOptional = scanOperator.getColumnMetaToColRefMap()
+                    .keySet().stream().filter(col -> col.getName().equals(partitionColumn.getName())).findFirst();
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:92]: Why not find in the map directly other than linear filter ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java [hunk 3]
```diff
@@ -62,85 +63,90 @@ public List<ColumnRefOperator> getOutputExpressions() {
...
-            return;
+            return resultRanges;
         }
-        List<Range<PartitionKey>> finalRanges = Lists.newArrayList();
+
         for (int i = 0; i < uptodatePartitionRanges.size(); i++) {
             Range<PartitionKey> currentRange = uptodatePartitionRanges.get(i);
             boolean merged = false;
-            for (int j = 0; j < finalRanges.size(); j++) {
+            for (int j = 0; j < resultRanges.size(); j++) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:137]: Is the `resultRanges` initialized ?

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 4]
```diff
@@ -853,6 +853,9 @@ private PartitionDiff getPartitionDiff(Expr partitionExpr, Column partitionColum
     }
 
     public Pair<Table, Column> getPartitionTableAndColumn() {
+        if (!(partitionInfo instanceof ExpressionRangePartitionInfo)) {
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:856]: no, not the same purpose. ExpressionRangePartitionInfo is used only by mv.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java [hunk 5]
```diff
@@ -62,85 +63,90 @@ public List<ColumnRefOperator> getOutputExpressions() {
...
+                                                Set<String> mvPartitionNamesToRefresh) {
+        Pair<Table, Column> partitionTableAndColumns = mv.getPartitionTableAndColumn();
+        if (partitionTableAndColumns == null) {
+            return false;
         }
-        ExpressionRangePartitionInfo exprPartitionInfo = (ExpressionRangePartitionInfo) partitionInfo;
-        if (partitionsToRefresh.isEmpty()) {
-            // all partitions are uptodate, do not add filter
-            return;
+        OlapTable partitionByTable = (OlapTable) partitionTableAndColumns.first;
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:73]: we can not get updated partitions for external table now. so external table is not supported.
you are right. I will add table type check here.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:104]: you mean latest?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java:137]: no, it is inited to Lists.newArrayList()

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 68 / 73 (93.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|42|47|89.36%|[50, 70, 74, 81, 131]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|3|3|100.00%|[]|

**Review by Youngwb — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**